### ISSUE TITLE
Attention layer does not accept output of previous layers in functional API

### ISSUE THREAD

**Issue by jorgenorena** [keras-team/keras#20318]
As an exercise to get acquainted with Keras, I want to train a simple model with attention to translate sentences.

I am not calling a tf function, only using Keras layers. But I get the following error:

A KerasTensor cannot be used as input to a TensorFlow function. A KerasTensor is a symbolic placeholder for a shape and dtype, used when constructing Keras Functional models or Keras Functions. You can only use it as input to a Keras layer or a Keras operation (from the namespaces `keras.layers` and `keras.operations`). [...]

Here is the code for the model using Keras' functional API:

```python
encoder_inputs = tf.keras.layers.Input(shape=[], dtype=tf.string)
decoder_inputs = tf.keras.layers.Input(shape=[], dtype=tf.string)

embed_size = 128
encoder_inputs_ids = text_vec_layer_en(encoder_inputs)
decoder_inputs_ids = text_vec_layer_es(decoder_inputs)
encoder_embedding_layer = tf.keras.layers.Embedding(vocab_size, embed_size, mask_zero=True)
decoder_embedding_layer = tf.keras.layers.Embedding(vocab_size, embed_size, mask_zero=True)
encoder_embeddings = encoder_embedding_layer(encoder_inputs_ids)
decoder_embeddings = decoder_embedding_layer(decoder_inputs_ids)

encoder = tf.keras.layers.LSTM(512, return_sequences=True, return_state=True)
encoder_outputs, *encoder_state = encoder(encoder_embeddings)

decoder = tf.keras.layers.LSTM(512, return_sequences=True)
decoder_outputs = decoder(decoder_embeddings, initial_state=encoder_state)

# Attention layer here!
# Problems getting it to work on Keras 3
attention_layer = tf.keras.layers.Attention()
attention_outputs = attention_layer([decoder_outputs, encoder_outputs])

output_layer = tf.keras.layers.Dense(vocab_size, activation="softmax")
Y_probas = output_layer(attention_outputs)
```

Expected behavior: The Keras attention layer accepts Keras tensor inputs. Or a more helpful error message is given.

Python version: 3.11.0
Tensorflow version: 2.17.0
Keras version: 3.4.1 (bundled with that Tensorflow version)

**Comment by mehtamansi29** [comment]
Hi @jorgenorena -

Thanks for reporting the issue. Based on code understand that you are trying to create model with attention for translate sentence. 
Here instead of using `tf.keras.layers.Attention` you can use `tf.keras.layers.MultiHeadAttention` with query,key and value for dot product. And then those attention output need to combine with decoder output and then create model using function API. 
Attached [gist](https://colab.sandbox.google.com/gist/mehtamansi29/71881cceb446df2d748404957ab8ea0b/20318-attention-layer-does-not-accept-output-of-previous-layers-in-functional-api.ipynb) for your reference here. 

`tf.keras.layers.Attention` is not fetching the input like this `attention_outputs = attention_layer([decoder_outputs, encoder_outputs])`. Here you can find more details about [attention](https://keras.io/api/layers/attention_layers/attention/) layer.

**Comment by jorgenorena** [comment]
Thanks @mehtamansi29 . This does give a working model.

However, my interest was not so much to build a model to translate, but rather to understand how the Keras interface works. Is the behavior of the Attention layer expected? If so, what is the logic? Or is this a bug?

Thanks for your help.

**Comment by THUdzc14** [comment]
Hi @jorgenorena! It's great to come across someone diving into the same ML textbook (assuming I guessed right :)). 

I ran into this exact error too. After spending an hour tinkering with the code and examining the example provided in the source code file of the Attention() class, I discovered a crucial point: when using the Attention layer, you can't use masking in conjunction with the embedding layer. Here's the code adjustment that resolved the issue for me:

```
# Do not set mask_zero=True when using Attention layer!
# encoder_embedding_layer = tf.keras.layers.Embedding(vocab_size, embed_size, mask_zero=True)
# decoder_embedding_layer = tf.keras.layers.Embedding(vocab_size, embed_size, mask_zero=True)
encoder_embedding_layer = tf.keras.layers.Embedding(vocab_size, embed_size)
decoder_embedding_layer = tf.keras.layers.Embedding(vocab_size, embed_size)
```

While this fixed the error, the underlying reason is still a puzzle to me, since I'm fairly new to TensorFlow and Keras, and I'm still on the learning curve. 

Another thing I noticed is that the parent class of Attention(), namely the BaseDenseAttention() class, clearly states in its documentation that it's designed for Dense or CNN networks and not intended for RNNs. However, the code seems to function just fine in our current context. It's quite the head-scratcher!

Wishing you all the best in your journey of exploring TensorFlow and ML. Hope this helps a bit, and if you have any further insights or thoughts, I'd love to hear them. :)
